interfaces.go

```diff
@@ -44,14 +44,19 @@ type DirectedGraph[NodeType any] interface {
 	Clone() DirectedGraph[NodeType]
 	// HasCycles performs cycle detection and returns true if the DirectedGraph has cycles.
 	HasCycles() bool
-	// PopReadyNodes returns of a list of all nodes that have finalized their status, whether
-	// resolved or unresolvable, and clears the list.
+	// PopReadyNodes returns of a list of all nodes that have no outstanding required dependencies,
+	// and clears the list.
 	// A node becomes ready to process when all of its AND dependencies and at least one of
 	// its OR dependencies are resolved.
+	// Note that nodes can be in any resolution state, since self-resolution is independent of its
+	// own dependency status. Some scenarios that can result in non-waiting nodes being included
+	// in the output include resolving them asynchronously by retrieving nodes independently of
+	// PopReadyNodes(), or due to a propagation of Unresolvable status to dependent nodes.
```

Comment:
The first sentence is excellent.  The second sentence is good, too, although it's not quite properly connected with the first sentence, because the first sentence talks about "outstanding required dependencies" while the second is talking about the concept of "ready to process".  I would be good if the two were better harmonized.  (Referencing "ready to process" is not bad, given the name of the function", but we need to tie that to the lack of outstanding dependencies first, and then describe what an outstanding dependency (or the lack thereof) is.

As for the rest, I would largely omit this.  Instead, I would say something like
> Note that the resolution state of a node is independent of its readiness and that the status values depending on the behavior of the calling code.

That is, I'm really hoping that, other than the primordial input node (which I'm hoping won't be left in the DAG for very long), there will never be a node on the ready list which is in the `Resolved` state when Arcaflow is using the DAG...so the "ready" nodes will always have a status of either `Waiting` or `Unresolvable`.

Reply:
I modified it to make the two sentences more harmonious, and I replaced the end with your recommended sentence.